File: viewer/src/shell.cpp
#include "shell.h"

#include "vpp/html.h"
#include "vpp/paint.h"

#include <cctype>
#include <cstdlib>

namespace {

// The shell page. ASCII only: the text renderer does not shape other
// scripts yet. Colours that depend on the theme are filled in by build().
const char* kShellHtml = R"(
<html><body>
<div id="bar">
  <button id="back" class="nav">&lt;</button>
  <button id="forward" class="nav">&gt;</button>
  <button id="reload" class="nav">R</button>
  <div id="address">
    <div id="address-text"></div>
  </div>
  <button id="minimize" class="win">_</button>
  <button id="maximize" class="win">[ ]</button>
  <button id="close" class="win close">X</button>
</div>
</body></html>
)";

const char* kShellCss = R"(
body { margin: 0; padding: 0; }
#bar { display: flex; align-items: center; gap: 6px; padding: 6px 8px; background-color: THEME; }
button.nav, button.win {
  padding: 4px 10px; font-size: 13px; border-radius: 6px;
  background-color: BUTTON; color: TEXT; font-weight: bold;
}
button.disabled { color: MUTED; }
button.close { background-color: #dc2626; color: white; }
#address { flex: 1; padding: 5px 10px; background-color: FIELD; border: 2px solid FIELD; border-radius: 6px; }
#address.focused { border: 2px solid #2563eb; }
#address-text { font-size: 13px; color: #18181b; }
#address.hidden { display: none; }
)";

std::string replaceAll(std::string s, const std::string& from, const std::string& to) {
    size_t pos = 0;
    while ((pos = s.find(from, pos)) != std::string::npos) {
        s.replace(pos, from.size(), to);
        pos += to.size();
    }
    return s;
}

// Relative luminance of a #rrggbb colour, 0..1; 1 (light) if unparseable.
float luminance(const std::string& hex) {
    if (hex.size() != 7 || hex[0] != '#') return 1.0f;
    const long v = std::strtol(hex.c_str() + 1, nullptr, 16);
    const float r = static_cast<float>((v >> 16) & 255) / 255.0f;
    const float g = static_cast<float>((v >> 8) & 255) / 255.0f;
    const float b = static_cast<float>(v & 255) / 255.0f;
    return 0.2126f * r + 0.7152f * g + 0.0722f * b;
}

std::string jsonString(const std::string& json, const std::string& key) {
    const size_t k = json.find("\"" + key + "\"");
    if (k == std::string::npos) return std::string();
    const size_t colon = json.find(':', k);
    if (colon == std::string::npos) return std::string();
    size_t i = colon + 1;
    while (i < json.size() && std::isspace(static_cast<unsigned char>(json[i]))) ++i;
    if (i < json.size() && json[i] == '"') {
        const size_t end = json.find('"', i + 1);
        return end == std::string::npos ? std::string() : json.substr(i + 1, end - i - 1);
    }
    size_t end = i;
    while (end < json.size() && (std::isalnum(static_cast<unsigned char>(json[end])) || json[end] == '.')) ++end;
    return json.substr(i, end - i);
}

} // namespace

WindowPrefs parseWindowPrefs(const std::string& json) {
    WindowPrefs prefs;
    if (json.empty()) return prefs;
    const std::string bar = jsonString(json, "addressBar");
    if (bar == "hidden" || bar == "shown") prefs.addressBar = bar;
    if (jsonString(json, "titleBar") == "false") prefs.titleBar = false;
    const std::string theme = jsonString(json, "theme");
    if (!theme.empty()) prefs.theme = theme;

    const std::string radius = jsonString(json, "cornerRadius");
    if (!radius.empty()) prefs.cornerRadius = std::strtof(radius.c_str(), nullptr);
    const std::string width = jsonString(json, "width");
    if (!width.empty()) prefs.width = std::atoi(width.c_str());
    const std::string height = jsonString(json, "height");
    if (height == "auto") prefs.autoHeight = true;
    else if (!height.empty()) prefs.height = std::atoi(height.c_str());

    const std::string resizable = jsonString(json, "resizable");
    if (resizable == "true") prefs.resizable = true;
    else if (resizable == "false") prefs.resizable = false;
    else prefs.resizable = !prefs.autoHeight;
    return prefs;
}

void Shell::build(const WindowPrefs& prefs) {
    prefs_ = prefs;
    const std::string theme = prefs.theme.empty() ? "#27272a" : prefs.theme;
    const bool dark = luminance(theme) < 0.5f;

    std::string css = kShellCss;
    css = replaceAll(css, "THEME", theme);
    css = replaceAll(css, "BUTTON", dark ? "rgba(255,255,255,0.14)" : "rgba(0,0,0,0.08)");
    css = replaceAll(css, "TEXT", dark ? "#ffffff" : "#18181b");
    css = replaceAll(css, "MUTED", dark ? "rgba(255,255,255,0.35)" : "rgba(0,0,0,0.3)");
    css = replaceAll(css, "FIELD", dark ? "#f4f4f5" : "#ffffff");

    document_ = vpp::parseHtml(kShellHtml);
    sheet_ = vpp::parseStyleSheet(css);
    root_.reset();

    if (prefs.addressBar == "hidden")
        if (vpp::Node* field = document_->findById("address")) field->setAttribute("class", "hidden");
    refreshAddressText();
}

void Shell::refreshAddressText() {
    if (!document_) return;
    if (vpp::Node* text = document_->findById("address-text")) {
        std::string shown = address.empty() && !focused ? "Enter a page address: a .vpp file or URL" : address;
        if (focused) shown += "|";
        text->setTextContent(shown);
    }
}

void Shell::setAddress(const std::string& text) {
    address = text;
    refreshAddressText();
}

void Shell::setNavigation(bool canGoBack, bool canGoForward) {
    if (!document_) return;
    if (vpp::Node* b = document_->findById("back")) b->setAttribute("class", canGoBack ? "nav" : "nav disabled");
    if (vpp::Node* f = document_->findById("forward")) f->setAttribute("class", canGoForward ? "nav" : "nav disabled");
}

void Shell::setFocused(bool on) {
    focused = on;
    if (!on) selectAll = false;
    if (!document_) return;
    if (vpp::Node* field = document_->findById("address")) {
        if (prefs_.addressBar != "hidden" || forced_) field->setAttribute("class", on ? "focused" : "");
    }
    refreshAddressText();
}

void Shell::layout(const vpp::FontSet& fonts, float viewportWidth, float scale) {
    if (!document_ || !visible()) {
        root_.reset();
        return;
    }
    // A forced reveal shows the address field even when the site hid it.
    if (vpp::Node* field = document_->findById("address")) {
        if (prefs_.addressBar == "hidden") field->setAttribute("class", forced_ ? (focused ? "focused" : "") : "hidden");
    }
    const vpp::LayoutContext ctx{fonts, scale, &sheet_};
    root_ = vpp::layoutDocument(*document_, ctx, viewportWidth);
}

void Shell::paint(vpp::Canvas& canvas, const vpp::FontSet& fonts, float scale) const {
    if (root_) vpp::paint(*root_, canvas, fonts, scale);
}

std::string Shell::controlAt(float x, float y) const {
    if (!root_) return std::string();
    const vpp::LayoutBox* box = vpp::hitTest(*root_, x, y);
    for (const vpp::Node* n = box ? box->node : nullptr; n; n = n->parent()) {
        if (!n->isElement()) continue;
        const std::string id = n->id();
        if (id == "back" || id == "forward" || id == "reload" || id == "minimize" || id == "maximize" ||
            id == "close" || id == "address")
            return id;
    }
    return std::string();
}

File: crates/vpp-viewer/src/shell.rs
//! The shell: the viewer's own bar above the page, drawn by the VPP engine.
//!
//! Back, forward, reload, the address field, and the window buttons. It is a
//! separate document with its own stylesheet, so pages cannot style it. A
//! site chooses only its colour and which parts show (`WindowPrefs`); the
//! colour is parsed first and written back as a colour, never as raw CSS.

use vpp_dom::{Document, NodeId, parse_html};
use vpp_layout::{LayoutBox, Rect, layout_document};
use vpp_paint::{DisplayItem, FontSet, build_display_list};
use vpp_style::values::color::parse_color;
use vpp_style::{Color, StyleSheet, parse_stylesheet};

use crate::window_prefs::WindowPrefs;

/// The shell page. ASCII labels: the text renderer does not shape other scripts yet.
const HTML: &str = r#"<html><body><div id="bar">
<button id="back" class="nav">&lt;</button>
<button id="forward" class="nav">&gt;</button>
<button id="reload" class="nav">R</button>
<div id="address"><div id="address-text"></div></div>
<button id="minimize" class="win">_</button>
<button id="maximize" class="win">[ ]</button>
<button id="close" class="win close">X</button>
</div></body></html>"#;

/// The stylesheet, with `THEME`, `BUTTON`, `TEXT`, `MUTED`, and `FIELD`
/// replaced by colours that suit the theme.
const CSS: &str = "
body { margin: 0; padding: 0; }
#bar { display: flex; align-items: center; gap: 6px; padding: 6px 8px; background-color: THEME; }
button.nav, button.win { padding: 4px 10px; font-size: 13px; border-radius: 6px;
  background-color: BUTTON; color: TEXT; font-weight: bold; }
button.disabled { color: MUTED; }
button.close { background-color: #dc2626; color: white; }
#address { flex: 1; padding: 5px 10px; background-color: FIELD; border: 2px solid FIELD; border-radius: 6px; }
#address.focused { border: 2px solid #2563eb; }
#address.hidden { display: none; }
#address-text { font-size: 13px; color: #18181b; }
";

const DEFAULT_THEME: Color = Color::rgb(39, 39, 42);
const PLACEHOLDER: &str = "Enter a page address: a .vpp file or URL";
/// The address text's size, as in `CSS`.
const TEXT_SIZE: f32 = 13.0;

/// A control in the shell.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Control {
    Back,
    Forward,
    Reload,
    Address,
    Minimize,
    Maximize,
    Close,
}

impl Control {
    fn from_id(id: &str) -> Option<Self> {
        Some(match id {
            "back" => Self::Back,
            "forward" => Self::Forward,
            "reload" => Self::Reload,
            "address" => Self::Address,
            "minimize" => Self::Minimize,
            "maximize" => Self::Maximize,
            "close" => Self::Close,
            _ => return None,
        })
    }
}

/// The shell and its state.
pub(crate) struct Shell {
    prefs: WindowPrefs,
    sheet: StyleSheet,
    document: Document,
    layout: Option<LayoutBox>,
    /// Shown in full whatever the site asked for: after Ctrl+L, or after
    /// following a link into another site. Esc ends it.
    pub(crate) forced: bool,
    /// The text in the address field.
    pub(crate) address: String,
    /// Whether the address field has the keyboard.
    pub(crate) focused: bool,
    /// Whether the next typed text replaces the whole address.
    pub(crate) select_all: bool,
    pub(crate) can_go_back: bool,
    pub(crate) can_go_forward: bool,
}

impl Shell {
    pub(crate) fn new() -> Self {
        let prefs = WindowPrefs::default();
        Self {
            sheet: stylesheet(&prefs),
            prefs,
            document: Document::new(),
            layout: None,
            forced: false,
            address: String::new(),
            focused: false,
            select_all: false,
            can_go_back: false,
            can_go_forward: false,
        }
    }

    pub(crate) fn prefs(&self) -> &WindowPrefs {
        &self.prefs
    }

    /// Uses a new page's preferences.
    pub(crate) fn set_prefs(&mut self, prefs: WindowPrefs) {
        self.sheet = stylesheet(&prefs);
        self.prefs = prefs;
    }

    pub(crate) fn visible(&self) -> bool {
        self.prefs.title_bar || self.forced
    }

    /// The bar's height in CSS pixels: 0 while hidden.
    pub(crate) fn height(&self) -> f32 {
        self.layout.as_ref().map_or(0.0, |root| root.frame.h)
    }

    pub(crate) fn contains(&self, x: f32, y: f32) -> bool {
        self.layout
            .as_ref()
            .is_some_and(|root| root.frame.contains(x, y))
    }

    /// The control under a point in CSS pixels, if any.
    pub(crate) fn control_at(&self, x: f32, y: f32) -> Option<Control> {
        let node = self.layout.as_ref()?.hit_test(x, y)?.node?;
        std::iter::successors(Some(node), |&n| self.document[n].parent())
            .find_map(|n| Control::from_id(self.document.element(n)?.id()?))
    }

    /// Builds the bar from the current state and lays it out `width` CSS pixels wide.
    pub(crate) fn lay_out(&mut self, fonts: &FontSet, width: f32) {
        if !self.visible() {
            self.layout = None;
            return;
        }
        let mut document = parse_html(HTML);
        let field_class = if !self.prefs.address_bar && !self.forced {
            "hidden"
        } else if self.focused {
            "focused"
        } else {
            ""
        };
        set_class(&mut document, "address", field_class);
        set_class(&mut document, "back", nav_class(self.can_go_back));
        set_class(&mut document, "forward", nav_class(self.can_go_forward));

        let text = if self.address.is_empty() && !self.focused {
            PLACEHOLDER.to_owned()
        } else {
            self.address.clone()
        };
        let caret = if self.focused { "|" } else { "" };
        // Set as text, never parsed: an address cannot inject markup.
        let node = document
            .find_by_id(document.root(), "address-text")
            .expect("the shell has an address field");
        let sheets = std::slice::from_ref(&self.sheet);

        // First with the field empty, to find how wide it is: a long address
        // must not push the window buttons out of the window.
        document.set_text_content(node, caret);
        let room = layout_document(&document, sheets, fonts, width)
            .and_then(|root| box_of(&root, node))
            .map_or(0.0, |field| field.w - fonts.regular.width(caret, TEXT_SIZE));
        document.set_text_content(node, format!("{}{caret}", fit(&text, room, fonts)));

        self.layout = layout_document(&document, sheets, fonts, width);
        self.document = document;
    }

    /// The border box of the shell element with `id`, for tests to click on.
    #[cfg(test)]
    pub(crate) fn box_of(&self, id: &str) -> Option<Rect> {
        let node = self.document.find_by_id(self.document.root(), id)?;
        box_of(self.layout.as_ref()?, node)
    }

    /// What the bar draws, in CSS pixels.
    pub(crate) fn display_list(&self) -> Vec<DisplayItem> {
        match &self.layout {
            Some(root) => build_display_list(root, &self.document),
            None => Vec::new(),
        }
    }
}

/// The border box of `node` in a laid-out shell.
fn box_of(b: &LayoutBox, node: NodeId) -> Option<Rect> {
    if b.node == Some(node) {
        return Some(b.frame);
    }
    b.children.iter().find_map(|c| box_of(c, node))
}

/// `text` if it fits in `room` CSS pixels, otherwise its end after "...":
/// the end of an address says most about where it leads.
fn fit(text: &str, room: f32, fonts: &FontSet) -> String {
    let font = &fonts.regular;
    if font.width(text, TEXT_SIZE) <= room {
        return text.to_owned();
    }
    let mut start = 0;
    for (i, _) in text.char_indices().skip(1) {
        start = i;
        if font.width(&format!("...{}", &text[i..]), TEXT_SIZE) <= room {
            break;
        }
    }
    format!("...{}", &text[start..])
}

fn nav_class(enabled: bool) -> &'static str {
    if enabled { "nav" } else { "nav disabled" }
}

fn set_class(document: &mut Document, id: &str, class: &str) {
    let node = document
        .find_by_id(document.root(), id)
        .expect("the shell has every control");
    if let Some(element) = document.element_mut(node) {
        element.set_attribute("class", class);
    }
}

/// The shell stylesheet in the site's colour, with text and buttons light
/// on a dark theme and dark on a light one.
fn stylesheet(prefs: &WindowPrefs) -> StyleSheet {
    let theme = prefs
        .theme
        .as_deref()
        .and_then(parse_color)
        .unwrap_or(DEFAULT_THEME);
    let dark = luminance(theme) < 0.5;
    let pick =
        |on_dark: &'static str, on_light: &'static str| if dark { on_dark } else { on_light };
    let css = CSS
        .replace("THEME", &css_color(theme))
        .replace("BUTTON", pick("rgba(255,255,255,0.14)", "rgba(0,0,0,0.08)"))
        .replace("MUTED", pick("rgba(255,255,255,0.35)", "rgba(0,0,0,0.3)"))
        .replace("FIELD", pick("#f4f4f5", "#ffffff"))
        .replace("TEXT", pick("#ffffff", "#18181b"));
    parse_stylesheet(&css)
}

fn css_color(c: Color) -> String {
    format!("rgba({},{},{},{})", c.r, c.g, c.b, f32::from(c.a) / 255.0)
}

/// Relative luminance, 0 (black) to 1 (white).
fn luminance(c: Color) -> f32 {
    (0.2126 * f32::from(c.r) + 0.7152 * f32::from(c.g) + 0.0722 * f32::from(c.b)) / 255.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_theme_is_a_colour_or_nothing() {
        // INTENTIONAL: hostile input. A theme that tries to add CSS rules must not reach the stylesheet.
        let prefs = WindowPrefs {
            theme: Some("red; } #address { display: none".into()),
            ..WindowPrefs::default()
        };
        assert_eq!(stylesheet(&prefs), stylesheet(&WindowPrefs::default()));

        let prefs = WindowPrefs {
            theme: Some("#2563EB".into()),
            ..WindowPrefs::default()
        };
        assert_ne!(stylesheet(&prefs), stylesheet(&WindowPrefs::default()));
    }

    #[test]
    fn dark_themes_get_light_text() {
        assert!(luminance(DEFAULT_THEME) < 0.5);
        assert!(luminance(Color::rgb(255, 255, 255)) > 0.5);
    }
}
